pkg/controller/kafka/cluster/setup.go

```diff
@@ -0,0 +1,172 @@
+/*
+Copyright 2021 The Crossplane Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cluster
+
+import (
+	"context"
+	"strings"
+
+	xpv1 "github.com/crossplane/crossplane-runtime/apis/common/v1"
+	"k8s.io/client-go/util/workqueue"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+
+	svcsdk "github.com/aws/aws-sdk-go/service/kafka"
+	"github.com/crossplane/crossplane-runtime/pkg/event"
+	"github.com/crossplane/crossplane-runtime/pkg/logging"
+	"github.com/crossplane/crossplane-runtime/pkg/meta"
+	"github.com/crossplane/crossplane-runtime/pkg/ratelimiter"
+	"github.com/crossplane/crossplane-runtime/pkg/reconciler/managed"
+	"github.com/crossplane/crossplane-runtime/pkg/resource"
+
+	svcapitypes "github.com/crossplane/provider-aws/apis/kafka/v1alpha1"
+	awsclients "github.com/crossplane/provider-aws/pkg/clients"
+)
+
+// SetupCluster adds a controller that reconciles Cluster.
+func SetupCluster(mgr ctrl.Manager, l logging.Logger, limiter workqueue.RateLimiter) error {
+	name := managed.ControllerName(svcapitypes.ClusterGroupKind)
+	opts := []option{
+		func(e *external) {
+			e.preObserve = preObserve
+			e.postObserve = postObserve
+			e.preDelete = preDelete
+			e.postDelete = postDelete
+			e.postCreate = postCreate
+			e.lateInitialize = LateInitialize
+		},
+	}
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		WithOptions(controller.Options{
+			RateLimiter: ratelimiter.NewDefaultManagedRateLimiter(limiter),
+		}).
+		For(&svcapitypes.Cluster{}).
+		Complete(managed.NewReconciler(mgr,
+			resource.ManagedKind(svcapitypes.ClusterGroupVersionKind),
+			managed.WithInitializers(managed.NewDefaultProviderConfig(mgr.GetClient())),
+			managed.WithExternalConnecter(&connector{kube: mgr.GetClient(), opts: opts}),
+			managed.WithLogger(l.WithValues("controller", name)),
+			managed.WithRecorder(event.NewAPIRecorder(mgr.GetEventRecorderFor(name)))))
+}
+
+func preDelete(_ context.Context, cr *svcapitypes.Cluster, obj *svcsdk.DeleteClusterInput) (bool, error) {
+	obj.ClusterArn = awsclients.String(meta.GetExternalName(cr))
+	return false, nil
+}
+
+func postDelete(_ context.Context, cr *svcapitypes.Cluster, obj *svcsdk.DeleteClusterOutput, err error) error {
+	if err != nil {
+		// skip: err.Error(): failed to delete Cluster: BadRequestException: You can't delete cluster in DELETING state.
+		if strings.Contains(err.Error(), "DELETING") {
+			return nil
+		}
```

Comment:
@haarchri this is likely a fine solution for now, but does AWS distinguish this as a specific type of error (e.g. is the status code unique). That may be a little more future-proof if this error message changed in the future.